hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/updatetokencustomfees/UpdateTokenCustomFeesDecoder.java

```diff
@@ -0,0 +1,201 @@
+                        .build())
+                .feeCollectorAccountId(addressIdConverter.convert(fee.get(FIXED_FEE_FEE_COLLECTOR)))
+                .build());
+    }
+
+    private Stream<CustomFee> decodeFractionalFees(
+            @NonNull final Tuple[] fractionalFees, @NonNull final AddressIdConverter addressIdConverter) {
+        return Arrays.stream(fractionalFees).map(fee -> CustomFee.newBuilder()
+                .fractionalFee(FractionalFee.newBuilder()
+                        .fractionalAmount(Fraction.newBuilder()
+                                .numerator(fee.get(FRACTIONAL_FEE_NUMERATOR))
+                                .denominator(fee.get(FRACTIONAL_FEE_DENOMINATOR))
+                                .build())
+                        .minimumAmount(fee.get(FRACTIONAL_FEE_MIN_AMOUNT))
+                        .maximumAmount(fee.get(FRACTIONAL_FEE_MAX_AMOUNT))
+                        .netOfTransfers(fee.get(FRACTIONAL_FEE_NET_OF_TRANSFERS))
+                        .build())
+                .feeCollectorAccountId(addressIdConverter.convert(fee.get(FRACTIONAL_FEE_FEE_COLLECTOR)))
+                .build());
+    }
+
+    private Stream<CustomFee> decodeRoyaltyFees(
+            @NonNull final Tuple[] royaltyFees, @NonNull final AddressIdConverter addressIdConverter) {
+        return Arrays.stream(royaltyFees).map(fee -> CustomFee.newBuilder()
+                .royaltyFee(RoyaltyFee.newBuilder()
+                        .exchangeValueFraction(Fraction.newBuilder()
+                                .numerator(fee.get(ROYALTY_FEE_NUMERATOR))
+                                .denominator(fee.get(ROYALTY_FEE_DENOMINATOR))
+                                .build())
+                        .fallbackFee(getFallbackFee(fee))
+                        .build())
+                .feeCollectorAccountId(addressIdConverter.convert(fee.get(ROYALTY_FEE_FEE_COLLECTOR)))
+                .build());
+    }
+
+    // In the solidity structure for Fixed Fees we have the property `bool useCurrentTokenForPayment` that is not
+    // present in the protobuf version.
+    // In order to determine the denominating token we need to check if the property is present.
+    private TokenID determineDenominatingToken(@NonNull final Tuple fee, @NonNull final TokenID tokenId) {
+        final boolean useCurrentToken = fee.get(FIXED_FEE_USE_CURRENT_TOKEN_FOR_PAYMENT);
+        return useCurrentToken
+                ? tokenId
+                : getIfPresent(fee.get(FIXED_FEE_TOKEN_ID), fee.get(FIXED_FEE_USE_HBARS_FOR_PAYMENT));
+    }
+
+    private @Nullable TokenID getIfPresent(@NonNull final Address address, final boolean useHbarsForPayment) {
+        final var tokenId = ConversionUtils.asTokenId(address);
+        return useHbarsForPayment || tokenId.equals(TokenID.DEFAULT) ? null : tokenId;
+    }
+
+    private @Nullable FixedFee getFallbackFee(@NonNull Tuple fee) {
+        final Address tokenAddress = fee.get(ROYALTY_FEE_TOKEN_ID);
```

Comment:
When puzzling this out I went and looked, of course, at the protobuf decl for `RoyaltyFee`.  And I found (as expected) that these token index constants are named just like the field.  And also the fields aren't really named in a self-explanatory way.  I myself would have preferred that amount/tokenId/useHbarsForPayment each had "fallbackFee" somewhere in their name.  Oh well. 

If you agree consider renaming these particular 3 token index constant to be something like "ROYALTY_FALLBACK_FEE_xxx".

Reply:
Renamed. It makes sense and kinda connects the protobuf with the solidity struct property.